api/pkg/s3/completemultipartupload.go

```diff
@@ -1,86 +1,90 @@
-	var client datastore.DataStoreAdapter
-	if objectMD == nil {
-		log.Errorf("No such object err\n")
-		response.WriteError(http.StatusInternalServerError, NoSuchObject.Error())
+	multipartUpload.UploadId = uploadId
 
+	completeMultipartBytes, err := ioutil.ReadAll(request.Request.Body)
+	if err != nil {
+		log.Errorln("unable to complete multipart upload when read request body. err:", err)
+		WriteErrorResponse(response, request, ErrInternalError)
+		return
 	}
-	client = getBackendByName(ctx, s, objectMD.Backend)
-	if client == nil {
-		response.WriteError(http.StatusInternalServerError, NoSuchBackend.Error())
+	complMultipartUpload := &model.CompleteMultipartUpload{}
+	if err = xml.Unmarshal(completeMultipartBytes, complMultipartUpload); err != nil {
+		log.Errorf("unable to parse complete multipart upload XML. data: %s, err: %v", string(completeMultipartBytes), err)
+		WriteErrorResponse(response, request, ErrMalformedXML)
 		return
 	}
 
-	resp, s3err := client.CompleteMultipartUpload(&multipartUpload, completeUpload, ctx)
-	log.Infof("resp is %v\n", resp)
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+	if len(complMultipartUpload.Parts) == 0 {
+		log.Errorf("unable to complete multipart upload. err: %v", errors.New("len(complMultipartUpload.Parts) == 0"))
+		WriteErrorResponse(response, request, ErrMalformedXML)
+		return
+	}
+	if !sort.IsSorted(model.CompletedParts(complMultipartUpload.Parts)) {
+		log.Errorf("unable to complete multipart upload. data: %+v, err: %v", complMultipartUpload.Parts, errors.New("part not sorted."))
+		WriteErrorResponse(response, request, ErrInvalidPartOrder)
 		return
 	}
 
-	// delete multipart upload record, if delete failed, it will be cleaned by lifecycle management
-	record := s3.MultipartUploadRecord{ObjectKey: objectKey, Bucket: bucketName, UploadId: UploadId}
-	s.s3Client.DeleteUploadRecord(context.Background(), &record)
+	// Complete parts.
+	var completeParts []*pb.CompletePart
+	for _, part := range complMultipartUpload.Parts {
+		part.ETag = strings.TrimPrefix(part.ETag, "\"")
+		part.ETag = strings.TrimSuffix(part.ETag, "\"")
+		completeParts = append(completeParts, &pb.CompletePart{PartNumber: part.PartNumber, ETag: part.ETag})
+	}
 
-	objectMD.Partions = nil
-	objectMD.LastModified = time.Now().Unix()
-	objectMD.InitFlag = "1"
-	//insert metadata
-	_, err := s.s3Client.CreateObject(ctx, objectMD)
-	if err != nil {
-		log.Errorf("err is %v\n", err)
-		response.WriteError(http.StatusInternalServerError, err)
+	ctx := common.InitCtxWithAuthInfo(request)
+	result, err := s.s3Client.CompleteMultipartUpload(ctx, &pb.CompleteMultipartRequest{
+		BucketName:    bucketName,
+		ObjectKey:     objectKey,
+		UploadId:      uploadId,
+		CompleteParts: completeParts,
+	})
+	if err != nil || result.ErrorCode != int32(ErrNoErr) {
+		log.Errorln("unable to init multipart. err:", err)
+		WriteErrorResponse(response, request, GetFinalError(err, result.ErrorCode))
```

Comment:
If err is not nil and result is nil, it will crash. Maybe HandleS3Error(...) can be used here.

Reply:
the same as above